### ISSUE TITLE
ADD bert-base-japanese-char

### ISSUE THREAD

**Issue by kyakuno** [axinc-ai/ailia-models#1154]
Mecabがなくても動きそうな日本語BERT。
https://huggingface.co/cl-tohoku/bert-base-japanese-char

**Comment by kyakuno** [comment]
とりあえずonnxに変換して、vocabだけでtokenizeしてみると良さそう。

**Comment by kyakuno** [comment]
v3があるのでこちらを使う
https://huggingface.co/cl-tohoku/bert-base-japanese-char-v3

**Comment by kyakuno** [comment]
```
cd sitepackages/transformers
python3 convert_graph_to_onnx.py --framework pt --model bert-base-japanese-char-v3 bert-base-japanese-char-v3.onnx
```
https://huggingface.co/transformers/v3.1.0/serialization.html

**Comment by kyakuno** [comment]
```
oading pipeline (model: cl-tohoku/bert-base-japanese-char-v3, tokenizer: cl-tohoku/bert-base-japanese-char-v3)
Some weights of the model checkpoint at cl-tohoku/bert-base-japanese-char-v3 were not used when initializing BertForMaskedLM: ['cls.seq_relationship.weight', 'cls.seq_relationship.bias']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
Error while converting the model: Folder /Users/kyakuno/Desktop/repos/ailia-models-ax/natural_language_processing/bert_maskedlm/work is not empty, aborting conversion
```

**Comment by kyakuno** [comment]
bert-base-japanese-char-whole-word-maskingの方であれば変換できる。

```
python3 convert_graph_to_onnx.py --framework pt --model cl-tohoku/bert-base-japanese-char-whole-word-masking ./work/bert-base-japanese-char-whole-word-masking.onnx --pipeline fill-mask
```

**Comment by kyakuno** [comment]
FP16だと精度が出ない。

**Comment by kyakuno** [comment]
```
kyakuno@KazukinoMacBook-Pro bert_maskedlm % python3 bert_maskedlm.py -a bert-base-japanese-char-whole-word-masking -e 2
 INFO arg_utils.py (13) : Start!
 INFO arg_utils.py (163) : env_id: 2
 INFO arg_utils.py (166) : MPSDNN-Apple M1 Max
 INFO model_utils.py (84) : ONNX file and Prototxt file are prepared!
 INFO bert_maskedlm.py (82) : Input text : 私[MASK]お金で動く。
 INFO bert_maskedlm.py (85) : Tokenized text : ['私', '[MASK]', 'お', '金', 'で', '動', 'く', '。']
 INFO bert_maskedlm.py (97) : Indexed tokens : [1007, 4, 87, 352, 17, 167, 66, 10]
 INFO bert_maskedlm.py (112) : Predicting...
 INFO bert_maskedlm.py (126) : Predictions : 
 INFO bert_maskedlm.py (130) : 0 [CLS]
 INFO bert_maskedlm.py (130) : 1 [MASK]
 INFO bert_maskedlm.py (130) : 2 [PAD]
 INFO bert_maskedlm.py (130) : 3 [UNK]
 INFO bert_maskedlm.py (130) : 4 [SEP]
 INFO bert_maskedlm.py (132) : Script finished successfully.
kyakuno@KazukinoMacBook-Pro bert_maskedlm % python3 bert_maskedlm.py -a bert-base-japanese-char-whole-word-masking -e 2 -e 0
 INFO arg_utils.py (13) : Start!
 INFO arg_utils.py (163) : env_id: 0
 INFO arg_utils.py (166) : CPU
 INFO model_utils.py (84) : ONNX file and Prototxt file are prepared!
 INFO bert_maskedlm.py (82) : Input text : 私[MASK]お金で動く。
 INFO bert_maskedlm.py (85) : Tokenized text : ['私', '[MASK]', 'お', '金', 'で', '動', 'く', '。']
 INFO bert_maskedlm.py (97) : Indexed tokens : [1007, 4, 87, 352, 17, 167, 66, 10]
 INFO bert_maskedlm.py (112) : Predicting...
 INFO bert_maskedlm.py (126) : Predictions : 
 INFO bert_maskedlm.py (130) : 0 の
 INFO bert_maskedlm.py (130) : 1 は
 INFO bert_maskedlm.py (130) : 2 [UNK]
 INFO bert_maskedlm.py (130) : 3 、
 INFO bert_maskedlm.py (130) : 4 (
 INFO bert_maskedlm.py (132) : Script finished successfully.
```

**Comment by kyakuno** [comment]
wordpieceモデルは32000tokenで、charモデルは4000token。

**Comment by kyakuno** [comment]
再変換してみる。

```
python3 convert_graph_to_onnx.py --framework pt --model cl-tohoku/bert-base-japanese-whole-word-masking ./work/bert-base-japanese-whole-word-masking.onnx --pipeline fill-mask
```

**Comment by kyakuno** [comment]
再変換すると、bert-base-japanese-whole-word-maskingもFP16で推論できなくなるので、エクスポータに問題がありそう。

**Comment by kyakuno** [comment]
昔のグラフは、最後の方のPowの前にCastが挟まっている。

**Comment by kyakuno** [comment]
古いPytorchを使ってみる。

```
pip3 install torch==1.7.1
python3 convert_graph_to_onnx.py --framework pt --model cl-tohoku/bert-base-japanese-char-whole-word-masking ./work/bert-base-japanese-char-whole-word-masking.onnx --pipeline fill-mask
```

|バージョン|出力|
|-----|-----|
|1.6.0 (昔変換したもの)|Castが挟まる|
|2.0.1 (これまでの流れで変換したもの)|Castが挟まらない|
|1.7.1 (今回の調査)|Castが挟まる|

**Comment by kyakuno** [comment]
1.7.1でexportするとCastが挟まるがうまく推論はできず。

**Comment by kyakuno** [comment]
Tensorをダンプしてみる。

```
    for i in range(0,ailia_model.get_blob_count()):
        try:
            data = ailia_model.get_blob_data(i)
            print("Idx", i, ailia_model.get_blob_name(i), data)
        except:
            continue
```

**Comment by kyakuno** [comment]
Idx 97のBlobName 300から[[[[nan nan nan nan nan nan]になる。

**Comment by kyakuno** [comment]
[normal.txt.zip](https://github.com/axinc-ai/ailia-models/files/11783939/normal.txt.zip)
[re.txt.zip](https://github.com/axinc-ai/ailia-models/files/11783940/re.txt.zip)

**Comment by kyakuno** [comment]
BlobName 300はattention_maskの出力とのAdd。

**Comment by kyakuno** [comment]
attention_maskからのSubの値が非常に大きくなっていてnanになっている。

```
Idx 6 211 -10000.0
Idx 7 212 [[[[-0. -0. -0. -0. -0. -0.]]]]
```

```
Idx 6 211 -3.4028235e+38
Idx 7 212 [[[[nan nan nan nan nan nan]]]]
```

**Comment by kyakuno** [comment]
-10000はFP16の範囲に入りそうなので、異常に大きくなっているのが不思議。

**Comment by kyakuno** [comment]
非常に小さい値を足す、というオペレーションなので、env=0の場合はNaNにならずに動く。

```
Idx 6 211 -3.4028235e+38
Idx 7 212 [[[[-0. -0. -0. -0. -0. -0.]]]]
```

**Comment by kyakuno** [comment]
下記のWarningが怪しい。

```
Loading pipeline (model: cl-tohoku/bert-base-japanese-whole-word-masking, tokenizer: cl-tohoku/bert-base-japanese-whole-word-masking)
Some weights of the model checkpoint at cl-tohoku/bert-base-japanese-whole-word-masking were not used when initializing BertForMaskedLM: ['cls.seq_relationship.weight', 'cls.seq_relationship.bias']
- This IS expected if you are initializing BertForMaskedLM from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification model from a BertForPreTraining model).
- This IS NOT expected if you are initializing BertForMaskedLM from the checkpoint of a model that you expect to be exactly identical (initializing a BertForSequenceClassification model from a BertForSequenceClassification model).
```

**Comment by kyakuno** [comment]
ここのbiasの値がおかしい。右が昔にエクスポートしたもの、左が新しくエクスポートしたもの。

![bert](https://github.com/axinc-ai/ailia-models/assets/38881041/2ab139a0-18d5-4ac5-ba4c-2f6481f51870)

**Comment by kyakuno** [comment]
get_extended_attention_maskで1e-4 = -10000に設定するのが期待値に見える。
https://github.com/huggingface/transformers/issues/14859

**Comment by kyakuno** [comment]
modeling_utils.pyのget_extended_attention_maskでは、dtypeを見て最小値を設定するようになっている。

```
        # Since attention_mask is 1.0 for positions we want to attend and 0.0 for
        # masked positions, this operation will create a tensor which is 0.0 for
        # positions we want to attend and the dtype's smallest value for masked positions.
        # Since we are adding it to the raw scores before the softmax, this is
        # effectively the same as removing these entirely.
        extended_attention_mask = extended_attention_mask.to(dtype=dtype)  # fp16 compatibility
        extended_attention_mask = (1.0 - extended_attention_mask) * torch.finfo(dtype).min
        return extended_attention_mask
```

**Comment by kyakuno** [comment]
2022年1月20日の下記のPRで、書き込まれる値が-1e4と-1e9から変更になっている。
https://github.com/huggingface/transformers/commit/d3cb28886ac68beba9a6646b422a4d727b056c0c

**Comment by kyakuno** [comment]
FineTuningをやりやすくするため？
https://github.com/huggingface/transformers/pull/17306

**Comment by kyakuno** [comment]
IMO, the -10e4 comes from the original Google implementation of BERT and we just copied it everywhere. However I've now seen a couple of issues related to this.
https://github.com/huggingface/transformers/issues/14859

**Comment by kyakuno** [comment]
Transformersのmdeling_utils.pyのget_extended_attention_maskを書き換えてエクスポートしてみる。

```
extended_attention_mask = (1.0 - extended_attention_mask) * 1e-4#torch.finfo(dtype).min
```

エクスポート。

```
python3 convert_graph_to_onnx.py --framework pt --model cl-tohoku/bert-base-japanese-char-whole-word-masking ./work/bert-base-japanese-char-whole-word-masking.onnx --pipeline fill-mask
```

**Comment by kyakuno** [comment]
正しく動いた。

Text : 私[MASK]お金で動く。

-e 0

```
 INFO bert_maskedlm.py (134) : Predictions : 
 INFO bert_maskedlm.py (138) : 0 の
 INFO bert_maskedlm.py (138) : 1 は
 INFO bert_maskedlm.py (138) : 2 [UNK]
 INFO bert_maskedlm.py (138) : 3 、
 INFO bert_maskedlm.py (138) : 4 (
 INFO bert_maskedlm.py (140) : Script finished successfully.
```

-e 2

```
 INFO bert_maskedlm.py (134) : Predictions : 
 INFO bert_maskedlm.py (138) : 0 の
 INFO bert_maskedlm.py (138) : 1 は
 INFO bert_maskedlm.py (138) : 2 [UNK]
 INFO bert_maskedlm.py (138) : 3 、
 INFO bert_maskedlm.py (138) : 4 (
 INFO bert_maskedlm.py (140) : Script finished successfully.
```
